assembly/src/parsers/io_ops.rs

```diff
@@ -255,70 +270,104 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
 
 // HELPER FUNCTIONS
 // ================================================================================================
+#[derive(PartialEq, Eq)]
+enum ParamType {
+    Decimal,
+    Hex,
+    HexAndDecimal,
+}
 
-fn parse_constants(op: &Token) -> Result<Vec<Felt>, ParsingError> {
+fn parse_constants(op: &Token) -> Result<Vec<u64>, ParsingError> {
     let mut constants = Vec::new();
     let param_idx = 1;
     let param_count = op.num_parts() - param_idx;
 
     // for multiple input parameters, parse & push each one onto the stack in order, then return
-    if param_count > 1 {
-        for param_idx in param_idx..=param_count {
-            let value = parse_element_param(op, param_idx)?;
-            constants.push(value);
-        }
-        return Ok(constants);
+    for param_idx in param_idx..=param_count {
+        let value = parse_hex_or_decimal_param(op, param_idx)?;
+        constants.push(value);
     }
+    Ok(constants)
+}
 
-    // for a single input, there could be one value or there could be a series of many hexadecimal
-    // values without separators
-    let param_str = op.parts()[param_idx];
-    if let Some(param_str) = param_str.strip_prefix("0x") {
-        // parse 1 or more hexadecimal values
-        let values = parse_hex_params(op, param_idx, param_str)?;
-        // push each value onto the stack in order
-        for &value in values.iter() {
-            constants.push(value);
-        }
-    } else {
-        // parse 1 decimal value and push it onto the stack
-        let value = parse_decimal_param(op, param_idx, param_str)?;
-        constants.push(value);
+fn parse_hex_params(op: &Token, param_str: &str) -> Result<Vec<u64>, ParsingError> {
+    let param_idx = 1;
+    // handle error cases where the hex string is poorly formed
+    if param_str.len() % HEX_CHUNK_SIZE != 0 {
+        // hex string doesn't contain a valid number of bytes
+        return Err(ParsingError::invalid_param(op, param_idx));
     }
 
-    Ok(constants)
+    // iterate over the multi-value hex string and parse each 8-byte chunk into a valid element
+    (0..param_str.len())
+        .step_by(HEX_CHUNK_SIZE)
+        .map(|i| parse_hex_param(op, param_idx, &param_str[i..i + HEX_CHUNK_SIZE]))
+        .collect()
 }
 
-fn parse_hex_params(
-    op: &Token,
-    param_idx: usize,
+fn parse_push_one(
+    token: &Token,
+    param_type: ParamType,
     param_str: &str,
-) -> Result<Vec<Felt>, ParsingError> {
-    // handle error cases where the hex string is poorly formed
-    let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
-        if param_str.len() % 2 != 0 {
-            // parameter string is not a valid hex representation
-            return Err(ParsingError::invalid_param(op, param_idx));
-        }
-        true
+) -> Result<Node, ParsingError> {
```

Comment:
I don't think we need to overcomplicate here with `param_type`. We can just use `radix` as in the example I provided. To ensure valid radix is used, we could add a `debug_assert`.